### ISSUE TITLE
How to correctly handle embedding tying?

### ISSUE THREAD

**Issue by xinghaow99** [facebookresearch/lingua#74]
Hi! Great repo and thanks for open-sourcing it. 

I'm trying to train a model with embedding tying, and I noticed that there is a `weight_tying` option in 'apps.main.transformer.LMTransformerArgs' and implemented it as follows:
```
if args.weight_tying:
    self.output.weight = self.embeddings.tok_embeddings.weight
```
There is an extra attribute `self.embeddings` so I removed it and got it running.

Now I found myself running into a new issue: the training loss seems to be decreasing smoothly but the val loss does not. I tried to load the consolidated state_dict and found that the embeddings were not tied. So I tried to manually tie the embeddings in the eval code after loading the state_dict but still got high eval loss. 

Any help are appreciated!

**Comment by wesbz** [comment]
Hi, there is indeed an issue we noticed.
We will push a PR ASAP but here are the changes you might want to do:
1. Replace the snippet you shared with the following
    ```python
    if args.weight_tying:
        self.output = TiedLinear(self.tok_embeddings)
    else:
        self.output = nn.Linear(
            args.dim,
            args.vocab_size,
            bias=False,
        )
    ```
2. Add the following class:
    ```python
    class TiedLinear(nn.Module):
        def __init__(self, tied_module: nn.Module) -> None:
            super().__init__()
            self.tied_module = tied_module
            if not hasattr(tied_module, "weight"):
                raise AttributeError(
                    "Provided module does not have attribute 'weight'. Please check your tied_module."
                )

        def __call__(self, x: torch.Tensor) -> torch.Tensor:
            return F.linear(x, self.tied_module.weight)

    ```

This should allow you to load a model with weight tying properly.

**Comment by mathuvu** [comment]
This should be solved now with PR #75 !

**Comment by xinghaow99** [comment]
> This should be solved now with PR [#75](https://github.com/facebookresearch/lingua/pull/75) !

Hi! Thanks for the fix.

I have another question: If one wants to train a model with shared weights—for instance, `tying layers.0.feed_forward.w1` and `layers.1.feed_forward.w1`, which might be distributed across different devices—what would be the right way to do this?

Thanks for the help!
